File: week-4/backspace-string-compare/backspace-string-compare.js
/**
 * @param {string} S
 * @param {string} T
 * @return {boolean}
 */
var backspaceCompare = function(S, T) {
    const stack1 = [];
    const stack2 = [];
    
    for (let i of S) {
        if (i === '#') {
            if (stack1.length) stack1.pop();
        } else
            stack1.push(i);
    }
    
    for (let i of T) {
        if (i === '#') {
            if (stack2.length) stack2.pop();
        } else
            stack2.push(i);
    }
    
    return stack1.join('') === stack2.join('');
};

File: week-4/backspace-string-compare/backspace-string-compare.py
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        stack1 = []
        stack2 = []
        for i in S:
            if i == '#':
                if len(stack1):
                    stack1.pop()
            else:
                stack1.append(i)
                
        for i in T:
            if i == '#':
                if len(stack2):
                    stack2.pop()
            else:
                stack2.append(i)
        
        return ''.join(stack1) == ''.join(stack2)
